### ISSUE TITLE
Implement multierror

### ISSUE THREAD

**Issue by husio** [iov-one/weave#529]
This is a proposal for implementing an `error` that represents a collection of error values. A naive implementation (not good enough) is

```go
// in out /errors package
type MultiError []*Error
func (MultiError) Error() String { return ... }
```

The current implementation of `Validate` method returns a single validation error. For example the [`SendMsg.Validate`](https://github.com/iov-one/weave/blob/4c171db0813396463978717193b753a8d9849088/x/cash/msg.go#L26-L46) method can fail for many reasons but it fails fast. Notice that because validation is delegated to `weave.Address` and `coin.Coin` returned error can be of many kinds.

In the current state when validation does not pass for more than one reason then it returns a random from the API perspective error. This is best visualized in an [empty message validation](https://github.com/iov-one/weave/blob/4c171db0813396463978717193b753a8d9849088/x/cash/msg_test.go#L40) - why it returns `ErrInvalidAmount` and not `ErrEmpty`, `ErrInvalidMsg` or `ErrInvalidInput`? There is no reason and no good answer to this.

Please consider whether adding such error implementation is not overcomplicating returned value.

## Acceptance criteria
- [ ] research on multi error implementations to get influence. Consider various API designs, for example [hashicorp implementation](https://godoc.org/github.com/hashicorp/go-multierror)
- [ ] multi error must work with existing [`errors.Error.Is`](https://github.com/iov-one/weave/blob/master/errors/errors.go#L124). Update documentation and tests,
- [ ] multi errror impementation must provide `Is` method,
- [ ] multi error must map well to the [ABCI error interface](https://github.com/iov-one/weave/blob/master/errors/abci.go#L28), 
- [ ] create a list of packages where the `Validate` method must be updated to return a multierror. For each of those subtasks (packages):
  - [ ]  method notation must stay the same,
  - [ ] update tests to test for all possible errors instead of just one.

**Comment by ethanfrey** [comment]
Interesting idea, I would like to see this in action.

I assume you would collect all errors in the validate method and then abort after that fails, or would you carry on and collect more errors? (It seems some of them, like Empty account, are only hit after validation is run).

**Comment by husio** [comment]
I will prepare an example change to show it in action. I am not fully convinced myself it this complication is worth it.

**Comment by husio** [comment]
I have prepared a proof of concept implementation.

Please have a look at a model validation in `x/paychan` https://github.com/iov-one/weave/blob/multierror_poc/x/paychan/model.go#L12-L43
`Validate()` method returns an `error` instance but because it is a collection, more than one error can be represented. In addition to a nice `Error()` message, `Error.Is` is matching more results which is shown in tests - 
https://github.com/iov-one/weave/blob/multierror_poc/x/paychan/model_test.go#L28-L51

A model is invalid for many reasons so both `ErrEmpty.Is` and `ErrInvalidModel.Is` returns true.

**Comment by ethanfrey** [comment]
Interesting. Very interesting

**Comment by ethanfrey** [comment]
Seems like a great model for the validation functions. Not only better results but the code flow looks nicer.

**Comment by ethanfrey** [comment]
Seems quite solid proposal.

Outstanding question on how to expose on abci interface, but that is not blocking to write the other code (great improvement for internal usage). And then we can make a separate PR for updating the abci Error 

- first iteration, just return first error in the batch, like now. later, do something smarter

**Comment by husio** [comment]
Initially the multi error type was called `Errors` and later I have renamed it to `ValidationError`. This is because of the ABCI Code. This allows me to [hardcode the code](https://github.com/iov-one/weave/blob/multierror_poc/errors/multi.go#L44). My thinking was that when a model or message is not valid, a single code, regardless the reason is enough. You can learn the details from the message.

The above assumption cannot be used for other cases. In my opinion multi error might not be useful for other cases so keeping the implementation specific might be acceptable.

**Comment by ruseinov** [comment]
I'm going to look into this and produce a spike

**Comment by ruseinov** [comment]
Actually, I have taken a look at @husio spike and looks similar to what I'd think of in terms of error.Is checking, collection all the errors and a custom abci code just to satisfy the interface. So I don't think we actually need another spike on this  one :/

**Comment by husio** [comment]
I think just go for it - implement a validation error.

What would be very helpful for testing is to test if given validation errors collection contains a certain error from a particular field. For example in [`PaymentChannel` model validation](https://github.com/iov-one/weave/blob/master/x/paychan/model.go#L36-L41) I would like to be able to test like this:

```go

model := PaymentChannel{...}
err := model.Validate() // Returns validation multierror
if !errors.ErrModel.Is(err) {
   t.Fatalf("unexpected error: %s", err)
}


// In addition to doing err.Is(...) I would like to ensure 
// that the validation failed because of certain fields. 
// Using `Is` method can only confirm error type, 
// not the field that caused it.

if err := errors.FieldError(err, "Total"); !errors.ErrModel.Is(err) {
    t.Fatalf("unexpected Total attribute error: %s", err)
}
if err := errors.FieldError(err, "Timeout"); !errors.ErrInput.Is(err) {
    t.Fatalf("unexpected Timeout attribute error: %s", err)
}
```

I am using non existing right now `errors.FieldError` to extract from a validation multierror an error that was reported for a given field/attribute/tag. This requires error to be reported together with this information, for example `errs := errs.With("Timeout", errors.ErrInput, "timeout in the past")`.
This pseudocode is written with a lot of assumptions in mind and I think it is up to the implementator to decide how to do this.
Bottom line - it would be awesome to test errors more precisely when returning more than one.

**Comment by ruseinov** [comment]
Yeah, I understand the benefits of having field errors - basically then we could test all the validation errors in one go instead of having to spin up (numField+1) test cases to try it all as they are being tested sequentially.

On the other hand I think that the whole idea of running validations after some of them have already failed could be a bit more expensive. The UX gets better though, because if you can reliably report these errors to the client - then there is no need to try several times. But afaik we have separate validation on the frontend and no good way to map errors to fields backend -> frontend, so I keep wondering why would we run all the validations instead of failing fast. 

Thoughts?

**Comment by husio** [comment]
If you are afraid there might a noticible slowdown, best is to measure it with a benchmark. Our validations are very basic and I don't expect noticible difference.

Just to get the feeling for the change, below is an example of how a model validation could look like in the new version. I think there is not much difference, both versions are in my opinion easy to read and write.

https://github.com/iov-one/weave/blob/master/x/paychan/model.go#L17-L49

```go
  // Validate ensures the payment channel is valid.                              
  func (pc *PaymentChannel) Validate() error {                                   
          var errs ValidationMultiError                                          
                                                                                 
          errs = errs.With("Metadata", pc.Metadata.Validate())                   
          errs = errs.With("Src", pc.Src.Validate())                             
                                                                                 
          if pc.SenderPubkey == nil {                                            
                  errs = errs.With("SenderPubkey", errors.Wrap(errors.ErrModel, "missing sender public key"))
          }                                                                      
                                                                                 
          errs = errs.With("Recipient", pc.Recipient.Validate())                 
                                                                                 
          if pc.Timeout < inThePast {                                            
                  errs = errs.With("Timeout", errors.Wrap(errors.ErrInput, "timeout is required"))
          } else {                                                               
                  errs = errs.With("Timeout", pc.Timeout.Validate())             
          }                                                                      
                                                                                 
          if pc.Total == nil || !pc.Total.IsPositive() {                         
                  errs = errs.With("Total", errors.Wrap(errors.ErrModel, "negative total"))
          }                                                                      
                                                                                 
          if len(pc.Memo) > 128 {                                                
                  errs = errs.With("Memo", errors.Wrap(errors.ErrModel, "memo too long"))
          }                                                                      
                                                                                 
          // Transfer value must not be greater than the Total value represented 
          // by the PaymentChannel.                                              
          if pc.Transferred == nil || !pc.Transferred.IsNonNegative() || pc.Transferred.Compare(*pc.Total) > 0 {
                  errs = errs.With("Transferred", errors.Wrap(errors.ErrModel, "invalid transferred value"))
          }                                                                      
                                                                                 
          if errs == nil {                                                       
                  return nil                                                     
          }                                                                      
          return errs                                                            
  }  
```

**Comment by ruseinov** [comment]
The slowdown would depend on what we are checking, I think just in general fail fast routine is better than accumulating errors to save some lines in testing code.

For actual production code I see no real benefit of using this, it's pretty much the same amount of code, but extra complexity. If there's not much difference - why implement this at all?

About ABCI code - I was thinking about this and it does make sense to return an ABCI code of the first error that got collected into multierr(validationError), because that would be a result if it failed fast.
If we still want to implement this - let's see if such behaviour would make sense.

**Comment by ruseinov** [comment]
Alright, I'm going to implement this and we can decide whether we want to integrate it or not
